src/unix/fs.c

```diff
@@ -375,6 +375,99 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
   return n;
 }
 
+static int uv__fs_opendir(uv_fs_t* req) {
+  uv_dir_t* dir;
+
+  dir = uv__malloc(sizeof(uv_dir_t));
+  if (dir == NULL)
+    goto error;
+
+  if (uv_mutex_init(&dir->dir_mutex) != 0)
+    goto error;
+
+  dir->dir = opendir(req->path);
+  if (dir->dir == NULL)
+    goto error;
+
+  req->ptr = dir;
+  return 0;
+
+error:
+  if (dir != NULL) {
+    uv_mutex_destroy(&dir->dir_mutex);
+    uv__free(dir);
+  }
+
+  req->ptr = NULL;
+  return -1;
+}
+
+static int uv__fs_readdir(uv_fs_t* req) {
+  uv_dir_t* dir;
+  uv_dirent_t* dirent;
+  struct dirent* res;
+  unsigned int dirent_idx;
+  unsigned int i;
+
+  dir = req->ptr;
+  dirent_idx = 0;
+  uv_mutex_lock(&dir->dir_mutex);
```

Comment:
@bnoordhuis How can you assume that the caller will not call readdir on the same dir with concurrency? Since the libuv threadpool for async FS ops will allow concurrency by default, and as I noted in the old PR https://github.com/libuv/libuv/pull/416#issuecomment-381521263 the [readdir(3)](http://man7.org/linux/man-pages/man3/readdir.3.html) is pretty clear on the matter, so I think a mutex is needed:

>In the current POSIX.1 specification (POSIX.1-2008), readdir() is not required to be thread-safe. However, in modern implementations (including the glibc implementation), concurrent calls to readdir() that specify different directory streams are thread-safe. In cases where multiple threads must read from the same directory stream, using readdir() with external synchronization is still preferable to the use of the deprecated readdir_r(3) function. It is expected that a future version of POSIX.1 will require that readdir() be thread-safe when concurrently employed on different directory streams.

Reply:
> how can you assume that the caller to libuv will not call uv_readdir on the same dir with concurrency?

Because the code currently isn't equipped to deal with that. The mutex guards the `dir->dir` handle but not the `dir->dirents` array. Two concurrent threads will overwrite each other's entries in that array.

> Since the libuv threadpool for async FS ops will allow concurrency by default then it's not in the control of the library

I expect we'll call that out in the documentation, that you should never have more than one request in flight.

That isn't so different from multiple `uv_fs_write()` requests for the same file descriptor, for example. We don't make any promises about correctness or order there, either.

Reply:
I think Ben is right. I've removed the mutex.